src/infra/src/table/distinct_values.rs

```diff
@@ -0,0 +1,218 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use sea_orm::{
+    entity::prelude::*, ColumnTrait, ConnectionTrait, EntityTrait, FromQueryResult, QueryFilter,
+    Schema, Set,
+};
+
+use super::get_lock;
+use crate::{
+    db::{connect_to_orm, ORM_CLIENT},
+    errors::{self, DbError, Error},
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, EnumIter, DeriveActiveEnum)]
+#[sea_orm(rs_type = "String", db_type = "String(StringLen::N(15))")]
+pub enum OriginType {
+    #[sea_orm(string_value = "stream")]
+    Stream,
+    #[sea_orm(string_value = "dashboard")]
+    Dashboard,
+    #[sea_orm(string_value = "report")]
+    Report,
+}
+
+/// Define the distinct fields table
+/// Primary key for this is composite of all fields, i.e.
+/// There will always be only one entry for specific origin-stream-field combination
+#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
+#[sea_orm(table_name = "distinct_value_fields")]
+pub struct Model {
+    #[sea_orm(primary_key, column_type = "Text")]
+    pub origin: OriginType,
+    #[sea_orm(primary_key, column_type = "Text")]
+    pub origin_id: String,
+    #[sea_orm(primary_key, column_type = "Text")]
+    pub org_name: String,
+    #[sea_orm(primary_key, column_type = "Text")]
+    pub stream_name: String,
+    #[sea_orm(primary_key, column_type = "Text")]
+    pub stream_type: String,
+    #[sea_orm(primary_key, column_type = "Text")]
+    pub field_name: String,
+}
+
+#[derive(Copy, Clone, Debug, EnumIter)]
+pub enum Relation {}
+
+impl RelationTrait for Relation {
+    fn def(&self) -> RelationDef {
+        panic!("No relations defined")
+    }
+}
+
+impl ActiveModelBehavior for ActiveModel {}
+
+#[derive(FromQueryResult, Debug)]
+pub struct DistinctFieldRecord {
+    pub origin: OriginType,
+    pub origin_id: String,
+    pub org_name: String,
+    pub stream_name: String,
+    pub stream_type: String,
+    pub field_name: String,
+}
+
+impl DistinctFieldRecord {
+    pub fn new(
+        origin: OriginType,
+        origin_id: &str,
+        org: &str,
+        stream: &str,
+        stream_type: String,
+        field: &str,
+    ) -> Self {
+        Self {
+            origin,
+            origin_id: origin_id.to_owned(),
+            org_name: org.to_owned(),
+            stream_name: stream.to_owned(),
+            stream_type,
+            field_name: field.to_owned(),
+        }
+    }
+}
+
+pub async fn init() -> Result<(), errors::Error> {
+    create_table().await?;
+    Ok(())
+}
+
+pub async fn create_table() -> Result<(), errors::Error> {
+    let client = ORM_CLIENT.get_or_init(connect_to_orm).await;
+    let builder = client.get_database_backend();
+
+    let schema = Schema::new(builder);
+    let create_table_stmt = schema
+        .create_table_from_entity(Entity)
+        .if_not_exists()
+        .take();
```

Comment:
It appears that here we are taking an "Entity-first" approach to the migration: defining the datatype for the entity and then generating the migration from it. The Sea ORM docs [recommend](https://www.sea-ql.org/SeaORM/docs/next/migration/writing-migration/#schema-first-or-entity-first) a "Schema-first" approach where one...
- writes a migration script
- executes the migration script locally to apply schema changes to the local DB
- then finally automatically generates the entity models in a separate file using the Sea ORM CLI tool.

I am personally biased towards the "Schema-first" approach as the docs suggest since it allows us to leverage the CLI tool to generate correct the complicated entity and relationship definitions. Ultimately however, I think the choice between "Schema-first" vs "Entity-first" is subjective and either could work. However, I think it'd probably be good for us to discuss the decision and decide on one approach to stick with.

Reply:
⚠️ Although I do think the decision between "Schema-first" versus "Entity-first" is subjective, there is one point that I think we should be very careful about, regardless of which of those options we choose.

Specifically, if we reference an model type from a migration, then I think it is important that 
1. that model type *not* be a public type (ie, it is private to the migration script)
2. that model type should never be changed after the migration script is published

The reasoning here is that if the public model type is treated as the current representation of the table, and we decide to change that model type in the future, then that change could break the migration script. I elaborated on this a bit more here: https://github.com/openobserve/openobserve/pull/5100#discussion_r1847368967

Reply:
Thee Sea ORM docs allude to some of the design considerations I'm describing here. Specifically [they mention](https://www.sea-ql.org/SeaORM/docs/next/migration/writing-migration/#schema-first-or-entity-first) that the "Entity-first" approach is ok...
> if you intend to never change the entity schema. Or, you can keep the original entity and embed a copy in the migration file.

Reply:
Hey, I'll read over this in detail tomorrow, but in short the reason I did in this way (entity first) because :
- I tried using sea orm migration, but ran into some issue with types (see above comment)
- So I had to write the migration in the old way of manually doing DB calls.
- In this way, it felt more natural to define an entity manually, and I was referencing the ShortUrl table/entity created , which I believe was also done in the same way.

Thus, I directly defined entity instead of a migration + cli generation. I agree that we should decide one approach to stick with just to have a common way of doing it, but I feel this kind of case would be an exception.

Also, regarding the private/public type - because this entity type is not tied with migration directly - the call to get dashboard from meta is plain sql call, and storing it in new table depends on the new table entity, I think it is ok in this case for that struct to be public.

Reply:
> Also, regarding the private/public type - because this entity type is not tied with migration directly

I might be misunderstanding you, but I believe that at the moment the `Entity` type and the migration are tied directly together. Specifically right here we generate the "create table" SQL from the `Entity` type 👇 
```
    let create_table_stmt = schema
        .create_table_from_entity(Entity)
        .if_not_exists()
        .take();
```
And the `Entity` type is derived from the `Model` type here...
```
#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "distinct_value_fields")]
pub struct Model {
   ...
}
```

So my main point is just that any changes to the the `Model` struct in the future would cause the behavior of the migration to change and be inconsistent before vs after such changes.